File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        open_stack = []
        for c in s:
            if c == "(" or c == "[" or c == "{":
                open_stack.append(c)
            else:
                if len(open_stack) == 0 or not self.matches(open_stack[-1], c):
                    return False
                op = open_stack.pop()
                
        return len(open_stack) == 0
    
    def matches(self, op: str, cl: str) -> bool:
        if op == "(": return cl == ")"
        if op == "[": return cl == "]"
        if op == "{": return cl == "}"
        return False
        

File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
const closedToOpen = new Map([
    ["}", "{"], [")", "("], ["]", "["],
])

var isValid = function(s) {
    let openQueue = []
    
    for (let i = 0; i < s.length; i++) {
        if (!closedToOpen.has(s[i])) {
            openQueue.push(s[i])
        } else {
            if (openQueue.length === 0) {
                return false
            }
            if (openQueue[openQueue.length-1] !== closedToOpen.get(s[i])) {
                return false
            }
            openQueue.pop()
        }
    }
    
    return openQueue.length === 0
};